### ISSUE TITLE
Proposal: change backend implementation

### ISSUE THREAD

**Issue by pnwx** [opencorelabs/fira#83]
tl;dr the hand-rolled Go and Protobuf implementation have got to go. We should replace them with Django.

## Background

First, some background. With the original vision of self-hosting being the dominant platform for use of Fira, I decided to use Go as the primary backend implementation language. The primary reason was around resource consumption: a single Go process is extremely efficient. It can use as little as just a couple MB of memory and happily spew back API responses for the frontend to consume. This even went so far as to make the Go process the root process in the Docker container: it would be responsible for starting and running the frontend process(es). I was even able to shoehorn a full Postgres instance as a subprocess of the Go process.

## The Pains

This was neat in concept, but painful in practice. There are several drawbacks:

1. Fragile/brittle coupling - a change in the frontend often requires a change in the backend Go process, kicking off a process of trial an error by the frontend engineer(s).
2. Bespoke in nature - i.e. not "boring" enough. This architecture would not be used in any larger production deployment, ever.
3. A Postgres subprocess will certainly fail in unpredictable ways - it was not meant to be hosted this way.

Beyond just the architecture of the process management, there are some other issues with using Go for this project. Some of those problems, in no particular order:

* A bespoke Go backend will be more difficult for random contributors to participate in, as they will have to learn the entire project architecture before making a change. Compared to using something like Rails or Django, where everything follows a similar convention, this will discourage would-be contributors.
* Too difficult to make small incremental progress on. Making changes like adding a new model and wiring it up to the HTTP API is a major undertaking in the current backend. Spreading the work across multiple evenings is quite difficult to maintain context on. This would be easier to do if there were a standardized framework like Rails or Django to build in - it takes away almost all of the decision making allowing me to focus on the task at hand.
* Unproven database access layer. While spent 1000s of hours working on Go at various jobs and projects, there is no "one way" to access the database like there is with common web frameworks. Furthermore, most web frameworks ORMs are sufficiently generic enough that they can easily support different SQL servers. The Go ORMs all require special handling (in your application code) to support multiple SQL implementations. 

There are many others, but these I feel are the most important

There are also downsides to switching away from Go as well, in my opinion, easily outweighed:

* Picking up a web framework in an interpreted language like Ruby or Python wil inevitably incur a switching cost - primarily the time to re-implement the frontend. Fortunately, there is very little to re-build as much of the time was spent inventing the original Go architecture.
* The big web frameworks like Laravel, Rails, Django, etc all use interpreted languages which are all slower and less memory efficient than Go. They also take _much_ longer to startup. I think this this just a cost of doing business. If we want to save a ton of time on the implementation, we will have to pay with some CPU cycles.
* No "cool" factor - I will be sad to see the original Go architecture get deleted. It was quite cool, and I was proud of it. But it's time to acknowledge the "sunk" cost here and let it go.

## Moving Forward

Moving forward, I would like to spend a couple days rewriting the backend in Django. This is a very familiar framework to me and many, many other developers (an order of magnitude or more than Go). Re-implementing the basic authentication layer (sign up/log in/JWTs) should take very little time at all since there are already great libraries for doing this in Django. That is basically all we've implemented thus far (and some App stuff, which is currently unused, and will go unused for some time). From there we can get on to the real work to be done here, which is providing the Fira frontend a institution connection and financial data API.

#### Hosting and Development

The hosting architecture will also change to be something more predictable and less bespoke. We'll still do a single-docker image, but it will be much simpler.

The root process on the image will become `supervisord`. This will manage the process lifecycle of the subprocesses:

* Nginx - which will host the static fira homepage assets and be a reverse proxy for both the Django and Next apps.
* Django - will run and host the API
* Next - will run the Fira frontend application

Supervisord is a very common process manager and should be much more reliable and less brittle than Go was. The config files can live directly in the repo. For development, we'll simply use different config files which will point to different configurations for nginx/Django/Next (for interactive development).

#### Self-hosting

Self-hosted instances will now default to using Sqlite in the "config" dir - basically like every other homelab service. This will be incredibly more reliable than the unproven "embedded" Postgres we're using before.

The only downside to all of this is self-hosters will have to dedicate more memory to the docker container. Likely to be around ~350mb (still mostly dominated by the NodeJS Next process).

**Comment by harryhexhash** [comment]
This sounds good to me. FWIW most of my time in a previous role was with Python/Django so I'd probably be more help on that side of the stack.

Curious why Python/Django vs rail/laraval/etc...? Also, wouldn't Go/Gin give a similar experience?

**Comment by pnwx** [comment]
IMO Django/Rails/Laraval are all roughly equivalent - and I know Python/Django the best of the three.

Re Go/Gin - having used both Django and Gin extensively - it's not even close. Gin is a micro framework for creating HTTP servers. The full-fat frameworks save you so much time

**Comment by harryhexhash** [comment]
Makes sense, I haven't used Gin much. Out of the other 3 I know Django the best, so thats a plus.
